api/client/proxy/transport/transportv1/client_test.go

```diff
@@ -0,0 +1,588 @@
+					clt := agent.NewClient(agentStreamRW)
+					keys, err := clt.List()
+					if err != nil {
+						return trail.ToGRPC(trace.Wrap(err))
+					}
+
+					if len(keys) != 1 {
+						return trail.ToGRPC(fmt.Errorf("expected to receive 1 key. got %v", len(keys)))
+					}
+
+					// send the key blob back via an ssh frame to alert the
+					// test that we finished listing keys
+					if err := server.Send(&proxyv1.ProxySSHResponse{
+						Details: nil,
+						Frame:   &proxyv1.ProxySSHResponse_Ssh{Ssh: &proxyv1.Frame{Payload: keys[0].Blob}},
+					}); err != nil {
+						return trail.ToGRPC(trace.Wrap(err))
+					}
+
+					return nil
+				},
+			},
+			assertion: func(t *testing.T, conn net.Conn, details *proxyv1.ClusterDetails, err error) {
+				require.NoError(t, err)
+				require.NotNil(t, conn)
+				require.True(t, details.FipsEnabled)
+
+				// write data via ssh frames
+				msg := []byte("hello")
+				n, err := conn.Write(msg)
+				require.NoError(t, err)
+				require.Equal(t, len(msg), n)
+
+				// read data via ssh frames
+				out := make([]byte, n)
+				n, err = conn.Read(out)
+				require.NoError(t, err)
+				require.Equal(t, len(msg), n)
+				require.Equal(t, msg, out)
+
+				// get the keys from our local keyring
+				keys, err := keyring.List()
+				require.NoError(t, err)
+				require.Len(t, keys, 1)
+
+				// the server performs a remote list of keys
+				// via ssh frames. to prevent the test from terminating
+				// before it can complete it will write the blob of the
+				// listed key back on the ssh frame. verify that the key
+				// it received matches the one from out local keyring.
+				out = make([]byte, len(keys[0].Blob))
+				n, err = conn.Read(out)
+				require.NoError(t, err)
+				require.Equal(t, len(keys[0].Blob), n)
+				require.Equal(t, keys[0].Blob, out)
+
+				// close the stream
+				require.NoError(t, conn.Close())
+			},
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+
+			pack := newServer(t, test.server)
+
+			conn, details, err := pack.Client.DialHost(context.Background(), test.cluster, test.target, nil, test.keyring)
+			test.assertion(t, conn, details, err)
+		})
+	}
+}
+
+// testPack used to test a [Client].
+type testPack struct {
+	Client *Client
+	Server proxyv1.ProxyServiceServer
+}
+
+// newServer creates a [grpc.Server] and registers the
+// provided [proxyv1.ProxyServiceServer] with it opens
+// an authenticated Client.
+func newServer(t *testing.T, srv proxyv1.ProxyServiceServer) testPack {
+	// gRPC testPack.
+	const bufSize = 100 // arbitrary
+	lis := bufconn.Listen(bufSize)
+	t.Cleanup(func() {
+		require.NoError(t, lis.Close())
+	})
+
+	s := grpc.NewServer()
+	t.Cleanup(func() {
+		s.GracefulStop()
+		s.Stop()
+	})
+
+	// Register service.
+	proxyv1.RegisterProxyServiceServer(s, srv)
+
+	// Start.
+	go func() {
+		if err := s.Serve(lis); err != nil {
+			panic(fmt.Sprintf("Serve returned err = %v", err))
```

Comment:
Is there an advantage to using a panic instead of `t.FailNow` or similar here ?

Reply:
I don't have particularly strong feelings (I'm not the world's biggest fan of `assert`/`require`) - but I do think that a large part of our test corpus currently uses `assert`/`require` and without any directive to suggest we are moving away from that, where it works, it's probably best to remain consistent with that.